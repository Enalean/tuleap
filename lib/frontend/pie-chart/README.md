# @tuleap/pie-chart

Makes it easier to create a d3 pie-chart.

## Usage

```typescript
import { StatisticsPieChart } from "@tuleap/pie-chart";

const pie_chart = new StatisticsPieChart({
    id : "pie-chart", // HTML id attribute
    prefix: "my-css-prefix", // CSS classes prefix
    general_prefix: "my-css-legend-prefix", // CSS class prefix for the legend
    data,
    width: 300, // in pixels
    height: 250, // in pixels
    radius: 200 // in pixels
});

pie_chart.init(); // create the pie chart

// redraw the pie chart (for example on window resize)
pie_chart.redraw({ width: 200, height: 200, radius: 150 });
```

## TLP color support

The chart supports assigning colors to chunks of data. To do so, in the `data` attribute when creating the chart, supply a "color":

typescript block (I cannot do it in this comment)
const data = [
  {
    key: "active", // The key will be used for CSS classes of that dataset.
    label: "Active users", // Translated text for the legend
    value: 67, // The number that will produce the pie slice
    color: 'tlp-success-color'
  },
  {
    key: "deleted",
    label: "Deleted users",
    value: 33,
    color: 'tlp-dimmed-color'
  }
];

Supported CSS class names are:
 - `tlp-danger-color`
 - `tlp-success-color`
 - `tlp-dimmed-color`
 - `tlp-info-color`
 - `tlp-warning-color`
 - `''` Empty string