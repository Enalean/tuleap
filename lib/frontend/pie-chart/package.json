{
  "author": "Enalean Team",
  "name": "@tuleap/pie-chart",
  "homepage": "https://tuleap.org",
  "license": "GPL-2.0-or-later",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "module": "dist/pie-chart.js",
  "main": "dist/pie-chart.umd.cjs",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "sass": "./dist/pie-chart.css",
      "import": "./dist/pie-chart.js",
      "require": "./dist/pie-chart.umd.cjs"
    }
  },
  "dependencies": {
    "d3-selection": "3.0.0",
    "d3-shape": "3.0.1",
    "d3-transition": "3.0.1"
  },
  "devDependencies": {
    "@tuleap/build-system-configurator": "workspace:*"
  },
  "scripts": {
    "build": "vite build",
    "watch": "vite build --watch --mode development --minify false"
  }
}
