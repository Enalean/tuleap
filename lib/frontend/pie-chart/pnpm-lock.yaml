lockfileVersion: '6.0'

settings:
  autoInstallPeers: false
  excludeLinksFromLockfile: false

dependencies:
  d3-selection:
    specifier: 3.0.0
    version: 3.0.0
  d3-shape:
    specifier: 3.0.1
    version: 3.0.1
  d3-transition:
    specifier: 3.0.1
    version: 3.0.1(d3-selection@3.0.0)

devDependencies:
  '@tuleap/build-system-configurator':
    specifier: workspace:*
    version: link:../build-system-configurator

packages:

  /d3-color@3.1.0:
    resolution: {integrity: sha512-zg/chbXyeBtMQ1LbD/WSoW2DpC3I0mpmPdW+ynRTj/x2DAWYrIY7qeZIHidozwV24m4iavr15lNwIwLxRmOxhA==}
    engines: {node: '>=12'}
    dev: false

  /d3-dispatch@3.0.1:
    resolution: {integrity: sha512-rzUyPU/S7rwUflMyLc1ETDeBj0NRuHKKAcvukozwhshr6g6c5d8zh4c2gQjY2bZ0dXeGLWc1PF174P2tVvKhfg==}
    engines: {node: '>=12'}
    dev: false

  /d3-ease@3.0.1:
    resolution: {integrity: sha512-wR/XK3D3XcLIZwpbvQwQ5fK+8Ykds1ip7A2Txe0yxncXSdq1L9skcG7blcedkOX+ZcgxGAmLX1FrRGbADwzi0w==}
    engines: {node: '>=12'}
    dev: false

  /d3-interpolate@3.0.1:
    resolution: {integrity: sha512-3bYs1rOD33uo8aqJfKP3JWPAibgw8Zm2+L9vBKEHJ2Rg+viTR7o5Mmv5mZcieN+FRYaAOWX5SJATX6k1PWz72g==}
    engines: {node: '>=12'}
    dependencies:
      d3-color: 3.1.0
    dev: false

  /d3-path@3.0.1:
    resolution: {integrity: sha512-gq6gZom9AFZby0YLduxT1qmrp4xpBA1YZr19OI717WIdKE2OM5ETq5qrHLb301IgxhLwcuxvGZVLeeWc/k1I6w==}
    engines: {node: '>=12'}
    dev: false

  /d3-selection@3.0.0:
    resolution: {integrity: sha512-fmTRWbNMmsmWq6xJV8D19U/gw/bwrHfNXxrIN+HfZgnzqTHp9jOmKMhsTUjXOJnZOdZY9Q28y4yebKzqDKlxlQ==}
    engines: {node: '>=12'}
    dev: false

  /d3-shape@3.0.1:
    resolution: {integrity: sha512-HNZNEQoDhuCrDWEc/BMbF/hKtzMZVoe64TvisFLDp2Iyj0UShB/E6/lBsLlJTfBMbYgftHj90cXJ0SEitlE6Xw==}
    engines: {node: '>=12'}
    dependencies:
      d3-path: 3.0.1
    dev: false

  /d3-timer@3.0.1:
    resolution: {integrity: sha512-ndfJ/JxxMd3nw31uyKoY2naivF+r29V+Lc0svZxe1JvvIRmi8hUsrMvdOwgS1o6uBHmiz91geQ0ylPP0aj1VUA==}
    engines: {node: '>=12'}
    dev: false

  /d3-transition@3.0.1(d3-selection@3.0.0):
    resolution: {integrity: sha512-ApKvfjsSR6tg06xrL434C0WydLr7JewBB3V+/39RMHsaXTOG0zmt/OAXeng5M5LBm0ojmxJrpomQVZ1aPvBL4w==}
    engines: {node: '>=12'}
    peerDependencies:
      d3-selection: 2 - 3
    dependencies:
      d3-color: 3.1.0
      d3-dispatch: 3.0.1
      d3-ease: 3.0.1
      d3-interpolate: 3.0.1
      d3-selection: 3.0.0
      d3-timer: 3.0.1
    dev: false
