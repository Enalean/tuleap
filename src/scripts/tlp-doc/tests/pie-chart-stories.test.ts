/*
 * Copyright (c) Enalean, 2025 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */
// @vitest-environment jsdom
import { expect, describe, it, beforeEach } from "vitest";
import type { PieChartArgs } from "../stories/graphs/pie-chart.stories";
import { default as PieChartMeta } from "../stories/graphs/pie-chart.stories";

describe("PieChart Storybook snapshot", () => {
    beforeEach(() => {
        document.body.innerHTML = "";
        const chart_element = PieChartMeta.render(PieChartMeta.args as PieChartArgs, {
            globals: {},
            parameters: {},
            args: {
                width: PieChartMeta.args.width,
                height: PieChartMeta.args.height,
                data: PieChartMeta.args.data,
                prefix: PieChartMeta.args.prefix,
                general_prefix: PieChartMeta.args.general_prefix,
            },
            viewMode: "story",
            loaded: undefined,
            abortSignal: undefined,
            canvasElement: undefined,
            hooks: undefined,
            originalStoryFn: undefined,
            step: undefined,
            context: undefined,
            canvas: undefined,
            mount: undefined,
            reporting: undefined,
            initialArgs: undefined,
            argTypes: undefined,
            componentId: "",
            title: "",
            kind: "",
            id: "",
            name: "",
            story: "",
            tags: [],
        }) as HTMLElement;
        document.body.appendChild(chart_element);
    });
    it("matches snapshot of pie-chart-container", () => {
        const container = document.body.querySelector(".pie-chart-container");
        expect(container.outerHTML).toMatchSnapshot();
    });
    it("matches snapshot of the legend", () => {
        const legend = document.body.querySelector(".my-legend-legend");
        expect(legend.outerHTML).toMatchSnapshot();
    });
});
