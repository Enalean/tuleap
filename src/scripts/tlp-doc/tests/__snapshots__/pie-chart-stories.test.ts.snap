// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PieChart Storybook snapshot > matches snapshot of pie-chart-container 1`] = `"<div class="pie-chart-container" id="normal-pie"><svg width="400" height="300"><g transform="translate(200,150)"><g class="my-pie-slice my-pie-slice-a tlp-swatch-flamingo-pink"><path class="my-pie-slice-path pie-chart-slice-path" d="M2.4492935982947065e-15,-40A40,40,0,0,1,9.572626571502322,38.83767269704208L6.381751047668215,25.891781798028056A26.666666666666668,26.666666666666668,0,0,0,1.6328623988631377e-15,-26.666666666666668Z"></path><text class="my-pie-slice-text pie-chart-slice-text" transform="translate(49.6354437049027,-6.026834012766159)" dy=".35em" style="text-anchor: start;">30</text></g><g class="my-pie-slice my-pie-slice-2 tlp-swatch-inca-silver"><path class="my-pie-slice-path pie-chart-slice-path" d="M9.572626571502322,38.83767269704208A40,40,0,0,1,-37.40064970741659,14.184195481701433L-24.93376647161106,9.456130321134289A26.666666666666668,26.666666666666668,0,0,0,6.381751047668215,25.891781798028056Z"></path><text class="my-pie-slice-text pie-chart-slice-text" transform="translate(-23.236158602188407,44.27280128266051)" dy=".35em" style="text-anchor: end;">15</text></g><g class="my-pie-slice my-pie-slice-3 tlp-swatch-chrome-silver"><path class="my-pie-slice-path pie-chart-slice-path" d="M-37.40064970741659,14.184195481701433A40,40,0,0,1,-7.347880794884118e-15,-40L-4.898587196589413e-15,-26.666666666666668A26.666666666666668,26.666666666666668,0,0,0,-24.93376647161106,9.456130321134289Z"></path><text class="my-pie-slice-text pie-chart-slice-text" transform="translate(-41.14919329468283,-28.40323733655778)" dy=".35em" style="text-anchor: end;">20</text></g></g></svg><div id="my-pie-legend"><ul class="my-legend-legend"><li class="my-pie-legend my-pie-legend-a" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-a pie-chart-span-legend-color tlp-swatch-flamingo-pink"></span><span class="my-pie-legend-text-span">Label 1</span></li><li class="my-pie-legend my-pie-legend-2" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-2 pie-chart-span-legend-color tlp-swatch-inca-silver"></span><span class="my-pie-legend-text-span">Label 2</span></li><li class="my-pie-legend my-pie-legend-3" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-3 pie-chart-span-legend-color tlp-swatch-chrome-silver"></span><span class="my-pie-legend-text-span">Label 3</span></li></ul></div></div>"`;

exports[`PieChart Storybook snapshot > matches snapshot of the legend 1`] = `"<ul class="my-legend-legend"><li class="my-pie-legend my-pie-legend-a" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-a pie-chart-span-legend-color tlp-swatch-flamingo-pink"></span><span class="my-pie-legend-text-span">Label 1</span></li><li class="my-pie-legend my-pie-legend-2" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-2 pie-chart-span-legend-color tlp-swatch-inca-silver"></span><span class="my-pie-legend-text-span">Label 2</span></li><li class="my-pie-legend my-pie-legend-3" style="width: 10px;"><span class="my-pie-legend-color-span my-pie-legend-color-3 pie-chart-span-legend-color tlp-swatch-chrome-silver"></span><span class="my-pie-legend-text-span">Label 3</span></li></ul>"`;
